### ISSUE TITLE
elpy can't find the python elpy module, even when added to PYTHONPATH

### ISSUE THREAD

**Issue by ricsonc** [jorgenschaefer/elpy#1588]
# Summary

I understand that the elpy python package is now packaged with elpy itself. However i've had issues with that (elpy 1.28 worked for me. 1.29.1 does not):

When I run elpy-config, it is unable to find the elpy module:

```
The Python interpreter could not find the elpy module. Make sure the
module is installed.

[run] ~/avpy -m pip install --user elpy
```

Interestingly, I also get this blank error message

```
There was an unexpected problem starting the RPC process. Please check
the following output to see if this makes sense to you. To me, it
doesn't.



```

Here is what I get when i run `(elpy-config--get-config)`

```
#s(hash-table size 65 test equal rehash-size 1.5 rehash-threshold 0.8 data ("emacs_version" "25.3.2" "python_rpc" "~/avpy" "python_rpc_executable" "/home/ricsonc/avpy" "python_interactive" "python" "python_interactive_executable" "/usr/bin/python" "virtual_env" nil "virtual_env_short" nil "error_output" "" ...))
```

my elpy-rpc-pythonpath looks correct `"/home/ricsonc/.emacs.d/elpa/elpy-20190416.1701/"`
In fact I also
1. launched python with that path in my pythonpath, confirming that I am able to import elpy in the interpreter when i do that
2. pointed `(setq elpy-rpc-python-command "~/avpy")`, where avpy is a bash script which runs my actual python prepending `PYTHONPATH='/home/ricsonc/.emacs.d/elpa/elpy-20190416.1701/'` -- it doesn't make a difference whether or not i do this
3. all other versions of the elpy module should be uninstalled, and i can't import elpy when that path is not in my PYTHONPATH, as expected.

# My configuration

## Result of `(elpy-config)`
```
Elpy Configuration

Virtualenv........: None
RPC Python........: /home/ricsonc/avpy
Interactive Python: python (/usr/bin/python)
Emacs.............: 25.3.2
Elpy..............: Not found (Python), 1.29.1 (Emacs Lisp)
Jedi..............: Not found
Rope..............: Not found
Autopep8..........: Not found
Yapf..............: Not found
Black.............: Not found
Syntax checker....: flake8 (/home/ricsonc/bin/flake8)

You have not activated a virtual env. While Elpy supports this, it is
often a good idea to work inside a virtual env. You can use M-x
pyvenv-activate or M-x pyvenv-workon to activate a virtual env.

The directory ~/.local/bin/ is not in your PATH. As there is no active
virtualenv, installing Python packages locally will place executables
in that directory, so Emacs won't find them. If you are missing some
commands, do add this directory to your PATH -- and then do
`elpy-rpc-restart'.

The Python interpreter could not find the elpy module. Make sure the
module is installed.

[run] ~/avpy -m pip install --user elpy

There was an unexpected problem starting the RPC process. Please check
the following output to see if this makes sense to you. To me, it
doesn't.



The jedi package is not available. Completion and code navigation will
not work.

[run] ~/avpy -m pip install --user jedi

The autopep8 package is not available. Commands using this will not
work.

[run] ~/avpy -m pip install --user autopep8

The yapf package is not available. Commands using this will not work.

[run] ~/avpy -m pip install --user yapf

The black package is not available. Commands using this will not work.

[run] ~/avpy -m pip install --user black
```

## Elpy configuration in my init.el
```
(setq elpy-rpc-python-command "~/avpy")
(setq elpy-project-root "~/av")
(elpy-enable)
```

**Comment by galaunay** [comment]
Thanks for the detailed report.

It seems that Elpy fails to parse the output from the RPC.
We had an issue (#1521) with json parsing that could be related.
Would you have a package related to json installed in Emacs (like json-mode for example) ?

In any case, it could be useful to have the output of the following python code (run with Elpy in the PYTHONPATH):

```Python
import json
import sys
import warnings

warnings.filterwarnings('ignore', category=FutureWarning)

try:
    import urllib2 as urllib
except ImportError:
    import urllib.request as urllib

from distutils.version import LooseVersion


def latest(package, version=None):
    try:
        response = urllib.urlopen('https://pypi.org/pypi/{package}/json'.format(package=package)).read()
        latest = json.loads(response)['info']['version']
        if version is None or LooseVersion(version) < LooseVersion(latest):
            return latest
        else:
            return None
    except:
        return None


config = {}
config['python_version'] = ('{major}.{minor}.{micro}'
                            .format(major=sys.version_info[0],
                                    minor=sys.version_info[1],
                                    micro=sys.version_info[2]))

try:
    import elpy
    config['elpy_version'] = elpy.__version__
except:
    config['elpy_version'] = None

try:
    import jedi
    if isinstance(jedi.__version__, tuple):
        config['jedi_version'] = '.'.join(str(x) for x in jedi.__version__)
    else:
        config['jedi_version'] = jedi.__version__
    config['jedi_latest'] = latest('jedi', config['jedi_version'])
except:
    config['jedi_version'] = None
    config['jedi_latest'] = latest('jedi')

try:
    import rope
    config['rope_version'] = rope.VERSION
    if sys.version_info[0] <= 2:
        config['rope_latest'] = latest('rope', config['rope_version'])
    else:
        config['rope_latest'] = latest('rope_py3k', config['rope_version'])
except:
    config['rope_version'] = None
    config['rope_latest'] = latest('rope')

try:
    import autopep8
    config['autopep8_version'] = autopep8.__version__
    config['autopep8_latest'] = latest('autopep8', config['autopep8_version'])
except:
    config['autopep8_version'] = None
    config['autopep8_latest'] = latest('autopep8')

try:
    import yapf
    config['yapf_version'] = yapf.__version__
    config['yapf_latest'] = latest('yapf', config['yapf_version'])
except:
    config['yapf_version'] = None
    config['yapf_latest'] = latest('yapf')

try:
    import black
    config['black_version'] = black.__version__
    config['black_latest'] = latest('black', config['black_version'])
except:
    config['black_version'] = None
    config['black_latest'] = latest('black')
json.dump(config, sys.stdout)
```
That's the bit of code executed to get the config parameters, it is apparently returning an empty string in your case, which is unexpected.

**Comment by ricsonc** [comment]
I figured out the issue. I was trying to use elpy with this rather unconventional python environment. I understand elpy tries to add its path to PYTHONPATH so that the elpy package shows up, but for some reason that wasn't working in this case. So I tried manually passing PYTHONPATH in by setting `elpy-rpc-command` to a bash script which runs `PYTHONPATH='/home/ricsonc/.emacs.d/elpa/elpy-20190416.1701/' my_weird_python`. When trying to run the above code snippet, I realized this doesn't make any sense -- it should be ``PYTHONPATH='/home/ricsonc/.emacs.d/elpa/elpy-20190416.1701/' my_weird_python "$@"` instead. That resolved the problem.

Thanks!

**Comment by galaunay** [comment]
I am glad you figured it out,
Thanks for the feedback.

**Comment by tensorfoo** [comment]
For people who stumble across this bug like me, I found this answer on a [stackflow:](https://emacs.stackexchange.com/questions/52652/elpy-doesnt-recognize-i-have-virtualenv-installed) 

`(setq elpy-rpc-virtualenv-path 'current)` 

add that to your .emacs and you should be good.

**Comment by Xunius** [comment]
> Thanks for the detailed report.
> 
> It seems that Elpy fails to parse the output from the RPC.
> We had an issue (#1521) with json parsing that could be related.
> Would you have a package related to json installed in Emacs (like json-mode for example) ?
> 
> In any case, it could be useful to have the output of the following python code (run with Elpy in the PYTHONPATH):
> 
> ```python
> import json
> import sys
> import warnings
> 
> warnings.filterwarnings('ignore', category=FutureWarning)
> 
> try:
>     import urllib2 as urllib
> except ImportError:
>     import urllib.request as urllib
> 
> from distutils.version import LooseVersion
> 
> 
> def latest(package, version=None):
>     try:
>         response = urllib.urlopen('https://pypi.org/pypi/{package}/json'.format(package=package)).read()
>         latest = json.loads(response)['info']['version']
>         if version is None or LooseVersion(version) < LooseVersion(latest):
>             return latest
>         else:
>             return None
>     except:
>         return None
> 
> 
> config = {}
> config['python_version'] = ('{major}.{minor}.{micro}'
>                             .format(major=sys.version_info[0],
>                                     minor=sys.version_info[1],
>                                     micro=sys.version_info[2]))
> 
> try:
>     import elpy
>     config['elpy_version'] = elpy.__version__
> except:
>     config['elpy_version'] = None
> 
> try:
>     import jedi
>     if isinstance(jedi.__version__, tuple):
>         config['jedi_version'] = '.'.join(str(x) for x in jedi.__version__)
>     else:
>         config['jedi_version'] = jedi.__version__
>     config['jedi_latest'] = latest('jedi', config['jedi_version'])
> except:
>     config['jedi_version'] = None
>     config['jedi_latest'] = latest('jedi')
> 
> try:
>     import rope
>     config['rope_version'] = rope.VERSION
>     if sys.version_info[0] <= 2:
>         config['rope_latest'] = latest('rope', config['rope_version'])
>     else:
>         config['rope_latest'] = latest('rope_py3k', config['rope_version'])
> except:
>     config['rope_version'] = None
>     config['rope_latest'] = latest('rope')
> 
> try:
>     import autopep8
>     config['autopep8_version'] = autopep8.__version__
>     config['autopep8_latest'] = latest('autopep8', config['autopep8_version'])
> except:
>     config['autopep8_version'] = None
>     config['autopep8_latest'] = latest('autopep8')
> 
> try:
>     import yapf
>     config['yapf_version'] = yapf.__version__
>     config['yapf_latest'] = latest('yapf', config['yapf_version'])
> except:
>     config['yapf_version'] = None
>     config['yapf_latest'] = latest('yapf')
> 
> try:
>     import black
>     config['black_version'] = black.__version__
>     config['black_latest'] = latest('black', config['black_version'])
> except:
>     config['black_version'] = None
>     config['black_latest'] = latest('black')
> json.dump(config, sys.stdout)
> ```
> 
> That's the bit of code executed to get the config parameters, it is apparently returning an empty string in your case, which is unexpected.

Is that python code still been used? I didn't find it in my `~/.emacs.d/elpa/elpy-20200202.2031/elpy` folder. I found that the `urllib.urlopen()` call just hangs forever, and [this SO post](https://stackoverflow.com/a/16773080/2005415) suggests an issue with the `urllib` module. I tried using `requests` instead and got results from the given URL.

I'm having multiple issue with `elpy`, please let me know I should open a new issue or not, because I found relevant pieces in different issues. Here are mine:

1. Running `elpy-config` hangs forever, I already mentioned in https://github.com/jorgenschaefer/elpy/issues/945#issuecomment-587210595. Could it be related to this `urllib` thing?
2. My `elpy-config` also reports the `elpy can't find the python elpy module` error. See below:

```
Elpy Configuration

Emacs.............: 26.3
Elpy..............: Not found (Python), 1.32.0 (Emacs Lisp)
Virtualenv........: None
Interactive Python: python 2.7.17 (/home/guangzhi/anaconda3/envs/cdat2/bin/python)
RPC virtualenv....: cdat2 (/home/guangzhi/anaconda3/envs/cdat2)
 Python...........: python nil (/home/guangzhi/anaconda3/envs/cdat2/bin/python)
 Jedi.............: Not found
 Rope.............: Not found
 Autopep8.........: Not found
 Yapf.............: Not found
 Black............: Not found
Syntax checker....: flake8 (/home/guangzhi/anaconda3/envs/cdat2/bin/flake8)

Warnings

You have not activated a virtual env. While Elpy supports this, it is
often a good idea to work inside a virtual env. You can use M-x
pyvenv-activate or M-x pyvenv-workon to activate a virtual env.

The Python interpreter could not find the elpy module. Please report
to: https://github.com/jorgenschaefer/elpy/issues/new.
```
Note that I've added `(setq elpy-rpc-virtualenv-path 'current)` so it's using an existing conda environment for RPC. 

I've added the `~/.emacs.d/elpa/elpy-20200202.2031/` to PYTHONPATH in my `.bashrc` file, so I can import `elpy` in my python running in the conda env. But, in `elpy-config` it still says `elpy can't find the python elpy module`.

I also tried commenting out anything conda related in `.bashrc` and re-install `elpy` all over again, letting it creates its own rpc env, still cant fix it. 

These env things are driving me crazy. Please help.

**Comment by galaunay** [comment]
> Is that python code still been used? 

Yes, in the variable `elpy-config--get-config`, defined in elpy.el:599. It is used to gather package versions and some relevant information about your python installation. It is just run so `elpy-config` can display those informations.

> this SO post suggests an issue with the urllib module.

You can try changing `urllib` to `requests` in `elpy-config--get-config` and see if it fixes your issue:

```elisp
(defvar elpy-config--get-config "import json
import sys
import warnings

warnings.filterwarnings('ignore', category=FutureWarning)

import requests

from distutils.version import LooseVersion


def latest(package, version=None):
    try:
        response = requests.get('https://pypi.org/pypi/{package}/json'.format(package=package)).content
        latest = json.loads(response)['info']['version']
        if version is None or LooseVersion(version) < LooseVersion(latest):
            return latest
        else:
            return None
    except:
        return None

...
```

> 1. Running elpy-config hangs forever, I already mentioned in #945 (comment). Could it be related to this urllib thing?

It could be, `elpy-config` is probably waiting for the code in `elpy-config--get-config` to finish...

> 2. My elpy-config also reports the elpy can't find the python elpy module error. See below:

This also could be due to the code in `elpy-config--get-config` failing to execute.
Let's try fixing that first :).

**Comment by Xunius** [comment]
@galaunay I found that `elpy-config--get-config` function, and edited the python snippet as you shown. Just to make sure, the changes are:
1. replacing the import of `urllib` with `request`,
2. changing the line in `latest()` from `response = urllib.urlopen('https://pypi.org/pypi/{package}/json'.format(package=package)).read()` to `response = requests.get('https://pypi.org/pypi/{package}/json'.format(package=package)).content`.

But that doesn't seem to fix my problem, `elpy-config` still hangs, and python still can't find the `elpy` module.

My system python can import `requests`, so does the conda env python.

**Comment by Xunius** [comment]
> @galaunay I found that `elpy-config--get-config` function, and edited the python snippet as you shown. Just to make sure, the changes are:
> 
>     1. replacing the import of `urllib` with `request`,
> 
>     2. changing the line in `latest()` from `response = urllib.urlopen('https://pypi.org/pypi/{package}/json'.format(package=package)).read()` to `response = requests.get('https://pypi.org/pypi/{package}/json'.format(package=package)).content`.
> 
> 
> But that doesn't seem to fix my problem, `elpy-config` still hangs, and python still can't find the `elpy` module.
> 
> My system python can import `requests`, so does the conda env python.

**CORRECT MYSELF**:
I noticed a `elpy.elc` file, I assume that's the compiled version of `elpy.el`, so I renamed that and then re-implemented the change from `urllib` to `requests`. Now, `elpy-config` finishes nearly instantly, but I got this:

```
Elpy Configuration

Emacs.............: 26.3
Elpy..............: Not found (Python), 1.32.0 (Emacs Lisp)
Virtualenv........: None
Interactive Python: python 3.8.1 (/usr/bin/python)
RPC virtualenv....: rpc-venv (/home/guangzhi/.emacs.d/elpy/rpc-venv)
 Python...........: python nil (/home/guangzhi/.emacs.d/elpy/rpc-venv/bin/python)
 Jedi.............: Not found
 Rope.............: Not found
 Autopep8.........: Not found
 Yapf.............: Not found
 Black............: Not found
Syntax checker....: flake8 (/usr/bin/flake8)

Warnings

You have not activated a virtual env. While Elpy supports this, it is
often a good idea to work inside a virtual env. You can use M-x
pyvenv-activate or M-x pyvenv-workon to activate a virtual env.

The Python interpreter could not find the elpy module. Please report
to: https://github.com/jorgenschaefer/elpy/issues/new.

There was an unexpected problem starting the RPC process. Please check
the following output to see if this makes sense to you. To me, it
doesn't.

Traceback (most recent call last):
  File "<string>", line 7, in <module>
ModuleNotFoundError: No module named 'requests'
```
So now the `requests` module is missing. I'm a bit confused, I can import that using my system python `/usr/bin/python`, as well as my conda python at `~/anaconda/envs/cdat2`. So which python is it using to import `requests`? I also tried activate the `~/anaconda/envs/cdat2` env or the `~/.emacs.d/elpy/rpc-venv` env before `elpy-config`, and it still reports `No module named 'requests`.

**Comment by gopar** [comment]
It looks like `elpy` is trying to import `requests` with the python installation that is in `/home/guangzhi/.emacs.d/elpy/rpc-venv/bin/python`. The quick fix here is to install requests there by doing `/home/guangzhi/.emacs.d/elpy/rpc-venv/bin/python -m pip install requests`. I'm afraid i'm a little out of touch and unfortunately not as active anymore to help debug this further.

Have you changed the value of `elpy-rpc-python-command`? Is it pointed to a valid python path? eg `/usr/bin/python` or just `python`?

**Comment by Xunius** [comment]
@gopar Thanks for the reply.
The `/home/guangzhi/.emacs.d/elpy/rpc-venv/bin/python -m pip install requests` fixed the missing `requests`.

But `elpy-config` still hangs, and python still can't find the `elpy` module. Running
```
/home/guangzhi/.emacs.d/elpy/rpc-venv/bin/python -c 'import elpy; print(elpy)'
```
actually prints the `elpy` module.

I didn't change `elpy-rpc-python-command`, its value is `python`, `which python` points to `/usr/bin/python`, which in my system is a symlink pointing to `/usr/bin/python3.8`.

**Comment by Xunius** [comment]
**UPDATE MYSELF AGAIN**:
I somehow managed to get it working this time, by changing this line
```
response = requests.get('https://pypi.org/pypi/{package}/json'.format(package=package)).content
```
to 
```
response = requests.get('https://pypi.org/pypi/{package}/json'.format(package=package), timeout=5).content
```
Now `elpy-config` finishes, although after about 30s of freezing, and python can find the `elpy` module.

Since this is stretching to far, let me give it a summary to help clear up confusion a bit.

## Original issue
1. running `M-x elpy-config RET` freezes, never finishes, and I had to `C-g` twice to abort.
2. after the aborting, it shows `Elpy..............: Not found (Python), 1.32.0 (Emacs Lisp)`, and a message down below:

```
Warnings

You have not activated a virtual env. While Elpy supports this, it is
often a good idea to work inside a virtual env. You can use M-x
pyvenv-activate or M-x pyvenv-workon to activate a virtual env.

The Python interpreter could not find the elpy module. Please report
to: https://github.com/jorgenschaefer/elpy/issues/new.
```
## My setup

* emacs: `GNU Emacs 26.3 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.10)
 of 2019-08-30`.
* OS: Manjaro linux
* Python: `/usr/bin/python` `is python3.8`. I also have anaconda installed.

## My fix

NOTE I'm not entirely sure about the internal workings so I'm only reporting what I did.

1. remove conda activate from my `.bashrc`.

Previously I have these in my `.bashrc`:
```
# >>> conda initialize >>>
# !! Contents within this block are managed by 'conda init' !!
__conda_setup="$('/home/guangzhi/anaconda3/bin/conda' 'shell.bash' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/home/guangzhi/anaconda3/etc/profile.d/conda.sh" ]; then
        . "/home/guangzhi/anaconda3/etc/profile.d/conda.sh"
    else
        export PATH="/home/guangzhi/anaconda3/bin:$PATH"
    fi
fi
unset __conda_setup
# <<< conda initialize <<<

conda activate cdat2
```
I commented all those lines.
NOTE that I tried leaving those lines there and re-implementing only the following steps, it also worked. So this may not be necessary.

2. Edit the `~/.emacs.d/elpa/elpy-20200202.2031/elpy.el` file, to change the `elpy-config--get-config` function to this:

```
(defvar elpy-config--get-config "import json
import sys
import warnings
import requests

warnings.filterwarnings('ignore', category=FutureWarning)

from distutils.version import LooseVersion

def latest(package, version=None):
    try:
        response = requests.get('https://pypi.org/pypi/{package}/json'.format(package=package), timeout=5).content
        latest = json.loads(response)['info']['version']
        if version is None or LooseVersion(version) < LooseVersion(latest):
            return latest
        else:
            return None
    except:
        return None


config = {}
...
```
In short, replace `urllib` with `requests`. NOTE that if you find a `elpy.elc` file exists, rename it or trash it.

3. Install `requests` by `~/.emacs.d/elpy/rpc-venv/bin/python -m pip install requests`, verify the installation by
```
~/.emacs.d/elpy/rpc-venv/bin/python -c 'import requests'
```
If nothing prints out, it is good. You can also do
```
~/.emacs.d/elpy/rpc-venv/bin/python -c 'import elpy'
```
to verify that the RPC python can find `elpy` module.

3. Breathe deeply, and launch emacs, then `M-x elpy-config`.
4. Wait a few minutes. 

A side note: I didn't customize the `elpy-rpc-python-commond` variable, its value is `python`.

**Comment by galaunay** [comment]
So if I understand correctly, you managed to fix the `Elpy..............: Not found (Python), 1.32.0 (Emacs Lisp)`, which means Elpy should work properly in python buffers (am I right on that ?).

But `elpy-config` still take quite a while to  display, which means `requests` may still fail to access  pypi and just timeout.

So we would need to use `request` instead of `urlib2` and do something to deal with timeouts, which I am happy with.

Could you just post your current `elpy-config` just to be sure everything is alright ?
(Sorry for asking you to go through the deep breathe and the waiting again :)).

**Comment by Xunius** [comment]
@galaunay 

> So if I understand correctly, you managed to fix the Elpy..............: Not found (Python), 1.32.0 (Emacs Lisp), which means Elpy should work properly in python buffers (am I right on that ?).

Yes.

> But elpy-config still take quite a while to display, which means requests may still fail to access pypi and just timeout.

Likely so.

> So we would need to use request instead of urlib2 and do something to deal with timeouts, which I am happy with.

I rarely code web-related things in python, and it could be just my internet. I found that urlib2 issue at this [SO post](https://stackoverflow.com/q/16772795/2005415) and applied that `request` fix accordingly.

Here is my `elpy-config`:

```
Elpy Configuration

Emacs.............: 26.3
Elpy..............: 1.32.0
Virtualenv........: None
Interactive Python: python 3.8.1 (/usr/bin/python)
RPC virtualenv....: rpc-venv (/home/guangzhi/.emacs.d/elpy/rpc-venv)
 Python...........: python 3.8.1 (/home/guangzhi/.emacs.d/elpy/rpc-venv/bin/python)
 Jedi.............: 0.16.0
 Rope.............: 0.16.0
 Autopep8.........: 1.5
 Yapf.............: 0.29.0
 Black............: 19.10b0
Syntax checker....: flake8 (/usr/bin/flake8)

Warnings

You have not activated a virtual env. While Elpy supports this, it is
often a good idea to work inside a virtual env. You can use M-x
pyvenv-activate or M-x pyvenv-workon to activate a virtual env.

Options

`Raised' text indicates buttons; type RET or click mouse-1 on a button
to invoke its action.  Invoke [+] to expand a group, and [-] to
collapse an expanded group.  Invoke the [Group], [Face], and [Option]
buttons below to edit that item in another window.
```

**Comment by galaunay** [comment]
I just merged a potential fix.
If you can upgrade to the latest version (from git), it shouldn't be a problem anymore.

Basically, It now checks if pypi is reachable in a decent amount of time and if not, doesn't retrieve infos from pypi.
